File: tests/compile-fail/ForeignNullableGenericCrossing.cs
// The .NET SURFACE the Kotlin case beside this file is refused against. A genuine C# API declaring a nullable
// value type INSIDE a generic argument — the one shape carrier-argument erasure (#86) leaves no Kotlin expression
// able to inhabit, because Kotlin's own `List<Int?>` is an `IReadOnlyList<object>` and constructing
// `System.Collections.Generic.List<Int?>()` from Kotlin erases its argument the same way.
//
// The two CONTROLS matter as much as the refused member: a DIRECT `int?` parameter is exactly what a Kotlin `Int?`
// is and must keep crossing, and a `Func<int?, string>` parameter is inhabited exactly too, because a delegate
// PARAMETER keeps its concrete `Nullable<V>` in the erasure. A predicate that called a delegate parameter an
// argument position refused this second one, which is why it is pinned here rather than left to the rule's prose.
using System;
using System.Collections.Generic;

namespace fgn
{
    public class Api
    {
        // REFUSED: `List<Nullable<Int32>>` is not a type any Kotlin expression has.
        public int CountPresent(List<int?> xs)
        {
            int n = 0;
            foreach (var x in xs) if (x.HasValue) n++;
            return n;
        }

        // REFUSED at the RETURN, which is a different channel: the node's own `ret` is the caller's KOTLIN view and
        // is erased as a Kotlin slot, so it reads `List<object>` and the crossing is invisible there. What the member
        // DECLARES has to be stamped beside the parameter vector or this case compiles and leaves a
        // `List<Nullable<Int32>>` on a stack typed as the unrelated Kotlin form.
        public List<int?> Make() { var l = new List<int?>(); l.Add(1); l.Add(null); return l; }

        // #354: an array element is the same reified-argument crossing. Kotlin's `Array<Int?>` is physically
        // `object[]`, not `Nullable<int>[]`, so neither direction has an inhabitable Kotlin value.
        public int CountPresentArray(int?[] xs)
        {
            int n = 0;
            foreach (var x in xs) if (x.HasValue) n++;
            return n;
        }

        public int?[] MakeArray() => new int?[] { 1, null };

        // REFUSED at a PROPERTY, which reaches the same stamp through the accessor.
        public List<int?> Items { get; } = new List<int?>();

        // REFUSED at a GENERIC method's return. These nodes take their parameter descriptor from the frontend and
        // never entered the resolution that establishes a declared return, so this family was unchecked entirely.
        public List<int?> MakeG<T>() { var l = new List<int?>(); l.Add(1); return l; }

        // REFUSED at a genuine public CLR FIELD, which is read through `ldfld` and so carries no parameter vector at
        // all — its declared type is the only thing that states the crossing.
        public List<int?> Storage = new List<int?>();

        // CONTROL: a direct `int?` IS a Kotlin `Int?`.
        public int OrElse(int? x, int d) => x ?? d;

        // CONTROL: a delegate PARAMETER keeps its `Nullable<Int32>` in Kotlin too.
        public string Describe(int? x, Func<int?, string> f) => f(x);
    }
}

File: tests/compile-fail/ForeignNullableGenericCrossing.kt
// #86 — a .NET member whose declaration NO Kotlin expression inhabits.
//
// Carrier-argument erasure makes a nullable value type `System.Object` in every reified argument, so a Kotlin
// `List<Int?>` is an `IReadOnlyList<object>` — and the `System.Collections.Generic.List<Int?>` built below erases
// the same way, to a `List<object>`. `List<Nullable<Int32>>` is therefore a type the language cannot produce, and
// `List<object>` does not inhabit it: two invariant reified generics with no conversion between them. The
// frontend type-checks the call (it sees `List<Int?>` on both sides), so the refusal has to come from the backend.
//
// A silent mis-typing is the one outcome that must not happen, so the crossing is refused at the call, naming the
// member and the slot. The CONTROLS below must keep compiling in the same file, which is what makes this a
// statement about the POSITION rather than about `Nullable` appearing anywhere in a foreign signature: a direct
// `int?` parameter and a `Func<int?, string>` parameter are both inhabited exactly.
import fgn.Api
import System.Collections.Generic.List as NetList

fun main() {
    val api = Api()
    println(api.OrElse(null, 7))                                // control: a direct int? parameter
    println(api.Describe(3) { v -> v?.toString() ?: "none" })    // control: a delegate parameter keeps Nullable<int32>
    // Built explicitly on the .NET type, which is the remedy the refusal must NOT offer: this construction
    // erases its own argument to `List<object>` exactly as a Kotlin `List<Int?>` does, so it reaches the same wall.
    println(api.CountPresent(NetList<Int?>()))                  // REFUSED: List<Nullable<Int32>> at a PARAMETER
}
